src/xenia/vfs/devices/stfs_container_device.h

```diff
@@ -24,144 +27,390 @@ namespace vfs {
+
+struct XContentAttributes {
+  uint8_t profile_transfer : 1;
+  uint8_t device_transfer : 1;
+  uint8_t move_only_transfer : 1;
+  uint8_t kinect_enabled : 1;
+  uint8_t disable_network_storage : 1;
+  uint8_t deep_link_supported : 1;
+  uint8_t reserved : 2;
+};
+static_assert_size(XContentAttributes, 1);
+
+XEPACKEDSTRUCT(XContentMetadata, {
+  static const uint32_t kThumbLength = 0x3D00;
+  static const uint32_t kThumbLengthV1 = 0x4000;
+
+  be<XContentType> content_type;
+  be<uint32_t> metadata_version;
+  be<uint64_t> content_size;
+  xex2_opt_execution_info execution_info;
+  uint8_t console_id[5];
+  be<uint64_t> profile_id;
   union {
     StfsVolumeDescriptor stfs_volume_descriptor;
-    SvodVolumeDescriptor svod_volume_descriptor;
+    SvodDeviceDescriptor svod_volume_descriptor;
+  };
+  be<uint32_t> data_file_count;
+  be<uint64_t> data_file_size;
+  be<XContentVolumeType> volume_type;
+  be<uint64_t> online_creator;
+  be<uint32_t> category;
+  uint8_t reserved2[0x20];
+  union {
+    XContentMediaData media_data;
+    XContentAvatarAssetData avatar_asset_data;
   };
-  uint32_t data_file_count;
-  uint64_t data_file_combined_size;
-  StfsDescriptorType descriptor_type;
   uint8_t device_id[0x14];
-  char16_t display_names[0x900 / 2];
-  char16_t display_descs[0x900 / 2];
-  char16_t publisher_name[0x80 / 2];
-  char16_t title_name[0x80 / 2];
-  uint8_t transfer_flags;
-  uint32_t thumbnail_image_size;
-  uint32_t title_thumbnail_image_size;
-  uint8_t thumbnail_image[0x4000];
-  uint8_t title_thumbnail_image[0x4000];
-
-  // Metadata v2 Fields
-  uint8_t series_id[0x10];
-  uint8_t season_id[0x10];
-  int16_t season_number;
-  int16_t episode_number;
-  char16_t additonal_display_names[0x300 / 2];
-  char16_t additional_display_descriptions[0x300 / 2];
+  union {
+    be<uint16_t> display_name_raw[9][128];
+    char16_t display_name_chars[9][128];
+  };
+  union {
+    be<uint16_t> description_raw[9][128];
+    char16_t description_chars[9][128];
+  };
+  union {
+    be<uint16_t> publisher_raw[64];
+    char16_t publisher_chars[64];
+  };
+  union {
+    be<uint16_t> title_name_raw[64];
+    char16_t title_name_chars[64];
+  };
+  union {
+    XContentAttributes bits;
+    uint8_t as_byte;
+  } flags;
+  be<uint32_t> thumbnail_size;
+  be<uint32_t> title_thumbnail_size;
+  uint8_t thumbnail[0x3D00];
+  union {
+    be<uint16_t> display_name_ex_raw[3][128];
+    char16_t display_name_ex_chars[3][128];
+  };
+  uint8_t title_thumbnail[0x3D00];
+  union {
+    be<uint16_t> description_ex_raw[3][128];
+    char16_t description_ex_chars[3][128];
+  };
+
+  std::u16string display_name(XLanguage language) const {
+    if (language >= XLanguage::kMaxLanguages) {
+      assert_always();
+      // no room for this lang, read from english slot..
+      language = XLanguage::kEnglish;
+    }
+
+    uint32_t lang_id = uint32_t(language) - 1;
+
+    const be<uint16_t>* str = 0;
+    if (lang_id >= 0 && lang_id < 9) {
```

Comment:
I'm not a big fan of magic numbers over there.
Same thing with few lines below

Reply:
Ah I did mean to do something about that but forgot to go over it, I guess since they're `metadata_version` related I could change them to `kMaxLanguagesV1`/`kMaxLanguagesV2` instead.